common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -540,6 +540,9 @@ public UTF8String trim() {
     // skip all of the space (0x20) in the right side
     int e = this.numBytes - 1;
     while (e > s && getByte(e) == 0x20) e--;
+    if (s == 0 && e == numBytes - 1) {
+      return this;
+    }
     return copyUTF8String(s, e);
```

Comment:
Can you avoid copying by changing `offset` if the string contains spaces at the left side, and changing `numBytes` for spaces at the right side?

Reply:
It looks non-safe if the string shared by others.

Reply:
I mean creating new instance of `UTF8String` by passing new value of `offset` or `numBytes`, and the same reference to `base`. Does `UTF8String` modify underlying base object somewhere in place?

Reply:
I see. My interpretation of `avoid copying` was `to update fieild in this object`. It functionally works well.

Such an optimization was done in `String.substring` in previous JDKs.   
Pros:  Avoid copying characters when `trim()`
Cons: Cannot free an object even if the  `UTF8String` object, which was referenced by the `trim`ed `UTF8String`, is dead under the case that `trim`ed `UTF8String` is live.

Reply:
What you're describing is kind of how the JVM implements "compressed strings". Interesting notes here on further optimizations in Java 9: https://www.baeldung.com/java-9-compact-string
One day we might revisit whether UTF8String is improving over String!